### DISCUSSION TITLE
Building for both Mac Intel & M1 using Github Actions

### DISCUSSION THREAD

**Discussion by Luthien-in-edhil** [wiverson/maven-jpackage-template#28] [General]
Hello again!

Just a question this time*: I picked up the dictionary app again after that ordeal of getting it to build on Windows & Mac last spring and because one of the people who want to test it has a new Mac M1. The Intel build fails to run on his machine, as expected.

Do you think that if could work to build both a M1- and an Intel-compatible Mac OS version of the app (next to the Windows- and Linux builds) using the Github Actions, by duplicating the Mac branch and using the M1 native JDK / JFX libraries in that new branch?

Thanks in advance! 
I see there have been some changes to the template since I last built it, so I'll first try things out in the existing configuration.

_(*) famous last words ... 😨_

**Comment by wiverson** [comment]
Hey! Hope you are well!

I just added some notes to the README for the project on M1 Mac stuff. AFAIK there are no M1 GitHub Action runners, and I don't think you can have an Intel version of the JDK assemble an M1 version via jlink.

The Intel version of the app *should* run via Rosetta. Might want to grab logs and see what's up with that.

I'm going to click the fancy button to convert this to a discussion. :)

**Comment by Luthien-in-edhil** [comment]
Thanks, I hope everything’s fine with you as well!

I’ll see if I can figure out the Rosetta way with my friend who has the M1 machine … it sounds like going back to the PowerPC days!

I haven’t looked into the exciting world of JFX / multiplatform for a while, but I suppose there have been no major new developments since April?

**Comment by wiverson** [comment]
RE Rosetta, yeah, no kidding. Especially with the same name, ha.

RE: news, nothing big. At this point I'm not sure how much interest there is in JavaFX vs just using https://www.formdev.com/flatlaf/ to build something in Swing. I know that people do find the modules stuff very confusing. I'm thinking of cloning off this project and making a parallel version that uses the FlatLaF and Swing - I know it would be a lot easier for folks to get started with.

Everyone in Java land is waiting for Java 17 + M1 to go final, I think it's Sept 14. I presume that JavaFX will also GA their 17 shortly thereafter.

**Reply by Luthien-in-edhil** [reply]
Modules confusing … 😵‍💫 …. well, the idea itself is great and clear enough, but in practice it creates a sort of classpath-to-the-nth-power dependency labyrinth. I remember you finding out that on Maven Central almost nothing had the right module info, so I can understand people’s reluctance to venture there! 

I still find it hard to understand (or accept?) how a situation developed where new developments don’t lead to more, but to fewer things being possible. Or, at least, attainable with a reasonable investment of effort and time. 

In that sense I can somewhat understand people reaching for Swing, because that at least _worked_. But it is such a pity because JavaFX is a thousand times better! 

So what it is - JavaFX not being used much, so nobody’s investing to really iron out those wrinkles? Or is there no demand for multi-platform + GUI? 
Whatever the reason is, people apparently don’t associate UI’s with Java, and I think that those who do think about OpenOffice right away. Or, worse, SoapUI - which extended the notion of “usability” considerably, in a downwards direction - it was quite an achievement!

I’m not sure modules are to blame, because the topic has never even been mentioned at work among my four Java developer colleagues, so it seems that there just as little awareness of modules in back-end Java.

**Comment by wiverson** [comment]
Oh yeah, nobody in server land worries about modules. It's more hassle to deal with tweaked or jlinked smaller VMs - just use the regular JDK and call it a day. It's only a factor if you are trying to build a shimmed down JVM. If you don't mind bundling the whole 50-100mb runtime you could skip everything about modules for your own apps.

RE: desktop Java (and for that matter, all x-plat desktop frameworks) as far as I can tell the market is just really really small. If you are a big company, you have dedicated teams building Swift for macOS and iOS, .NET for Windows, and Kotlin in Android. Then you probably have a Java team handling the backend with REST APIs and call it a day.

At the other end, it's pretty hard to argue with the web as a platform for both desktop and mobile users. Cake to deploy and test. The new flex stuff in CSS plus Bootstrap or Tailwind makes it really easy to build responsive UIs even without a lot of breakpoint fussing. Don't have to worry about giving up 30% to an app store vendor.

Also, games are all pretty much built with stuff like Unity or (more rarely) Java frameworks like libGDX. So if you want to build games, you are probably looking at one of those, not JavaFX or Swing.

TBH I see the browser runtimes getting a lot better faster than either Swing or JavaFX for most stuff. I see stuff like full 3D modeling tools, sound and video editors in HTML+JS. I dunno - crazy times.

**Reply by Luthien-in-edhil** [reply]
A web app should work indeed, though we want to add the option to add / use custom data, so that would require online persistence (doable) and a need for the user to log in and maintain an account (more complex). 

Speaking of game engines - I’ve been playing around with both Unity and Unreal engine lately, and their multi-platform magic is impressive. I can just build a Windows executable directly on the Mac version of the Unity editor - simply choose the option and press “build”. It rumbles for a while and spits out a perfectly fine Windows app. And that for all major desktop and mobile platforms. 

Which led me to the absolutely crazy thought of “would it be possible to package a regular app plus a database inside a Unity shell, so that it would work like a normal app?” 
Probably, yes. 
I’m not yet seriously considering it, but it’s weird enough for that!

**Reply by wiverson** [reply]
Oh, I spent a lot of time with Unity lately. Check out https://blazesky.com/ - I built all of that in Unity. Before that I was doing experiments with libGDX. I built it to support 8 languages, had it translated with dynamic switching between the languages.

I would suggest looking at Godot instead if you just want to build a light-weight UI on top of a database. But you absolutely could build a pretty slick UI in Unity and embed a database (e.g. SQLlite http://www.mwmdragon.com/2019/07/15/sqlite-in-unity/).

You will also want to look at https://docs.unity3d.com/Manual/com.unity.textmeshpro.html to see how to include custom fonts.

That said, I think you might find Godot easier to work with with - it's certainly a lot lighter and IMHO easier to work with for a lot of this stuff.

SQL Lite & Godot
https://github.com/2shady4u/godot-sqlite
Dynamic Fonts
https://docs.godotengine.org/en/stable/classes/class_dynamicfont.html
Rider/JetBrains integration
https://plugins.jetbrains.com/plugin/13882-godot-support
Switch Godot to use Darcula theme
https://github.com/godotengine/godot-syntax-themes
Use Supabase for Postgres in the cloud, really slick stuff
https://github.com/fenix-hub/godot-engine.supabase

That last one (using Supabase) might be a really good fit. You could use the free tier of Supabase to get a nice Postgres, and use their JWT auth + storage stuff to allow for users to have accounts and store custom data.

**Reply by wiverson** [reply]
Put my list of Godot tips here https://gist.github.com/wiverson/d2cb2b933242b5c57c41a4f90564cc55

**Reply by wiverson** [reply]
https://docs.godotengine.org/en/stable/getting_started/workflow/export/exporting_for_web.html

**Reply by abulka** [reply]
I'm really liking this project and the whole JavaFX / Scenebuilder approach. I think distributing apps easily is a major turn-off for developers looking at the Java ecosystem, and this template project seems to solve that problem! I was able to run my .dmg generated with this template on a brand new Mac, no JDK present - and it ran!

Speaking of alternative languages and UI frameworks - another alternative is Flutter from Google which is cross-platform, fast, a nice language (Dart) and a lovely developer experience - except the code driven nested widget approach gets pretty tiring to write, when all I want is something like Scenebuilder.

Coming from a Python background, I've used wxPython to build cross-platform apps, including deployment as a Snap in Ubuntu (see my project [Pynsource](https://github.com/abulka/pynsource)). However wxPython is getting pretty old and has only one maintainer, so I wouldn't build my next project in it.

I took a look at .NET 6 which is promising cross-platform UI approach these days, however their [Maui UI project](https://devblogs.microsoft.com/dotnet/update-on-dotnet-maui/) involves fiddling with XML to build GUIs - nothing like Scenebuilder is available.

Finally, there is Electron which uses web technologies - which has all the benefits you mention. The only problem there is you have to pick a UI toolkit and there are so many incomplete ones that it is difficult to choose - but there certainly are many individual components to choose from. I wish there was a dominant, free UI component set where everything had been thought out. If you use a framework like Vue then you get an advanced UI model/view and state management approach which is cutting edge and helps avoid spaghetti UI coding. The tooling for Electron + Vue is a learning curve - you can't just open an IntelliJ project and go!

I miss the days of Delphi where you open a project, drag out a UI, double click to add some handler code, Run.  Then when happy, save as .EXE and you are done. All we need these days is a version of that (albeit using a better language), and more deployment options - save as .EXE, .APP, Snap, Web App, etc.

**Reply by wiverson** [reply]
I've been thinking about this a lot over the last few months. I'm starting to think that the ideal framework might be a set of wrappers for https://github.com/webview/webview and [sqlite](https://sqlite.org/index.html). Then folks could write their logic and desktop handling features in Java and use something like [Pinegrow](https://pinegrow.com/) to build the UI in HTML. I've had several people ask for a "Spring Boot" like experience for building a Java desktop app, and this would be pretty darn close.

A lot of it comes down to all sorts of questions around target platform and experience. Right now TBH if I had to build an ordinary desktop+mobile app I'd probably be looking at Flutter. That said, I'm actually working on a desktop app that needs a lot of graphics access and am finding Unity to work pretty well. Interesting times.
